File: 0704. Binary Search.py
#sorted, unique, O(log n) --> binary search

#Python2
class Solution(object):
    def search(self, nums, target):
        left, right = 0, len(nums)-1

        while left <= right:
            middle = left + (right-left) // 2

            if nums[middle] > target:
                right = middle - 1 #target在左區間
            elif nums[middle] < target:
                left = middle + 1 #target在右區間
            else:
                return middle
        return -1 #未找到目標值

#Python3 1/3/24
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1  #定義區間為左閉右閉

        while left <= right:
            middle = (left + right) // 2
            if target < nums[middle]:
                right = middle - 1      #因為有算middle本身
            elif target > nums[middle]:
                left = middle + 1
            else:
                return middle
        return - 1

File: Java/1. Array/0704. Binary Search.java
class Solution {
    public int search(int[] nums, int target) {
        // ascending order (= ordered)
        // not exist  return - 1
        // TC: O(log n)
        int left = 0;
        int right = nums.length - 1;

        // The if: 1 2 3 (4) 5 6 7 -> tar: 2
        // The elif: 1 2 3 (4) 5 6 7 -> tar: 5

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > target) { // tar in the left
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] == target) {
                return mid;
            }
        }
        return - 1;
    }
}